jerry-core/ecma/builtin-objects/ecma-builtin-regexp-prototype.c

```diff
@@ -627,76 +501,112 @@ inline bool JERRY_ATTR_ALWAYS_INLINE
+                                                                                        *   passed to routine */
+                                                ecma_length_t arguments_number) /**< length of arguments' list */
 {
-  return ecma_regexp_replace_helper (this_arg, string_arg, replace_arg);
-} /* ecma_builtin_regexp_prototype_symbol_replace */
+  JERRY_UNUSED (arguments_number);
 
-/**
- * The RegExp.prototype object's '@@search' routine
- *
- * See also:
- *          ECMA-262 v6.0, 21.2.5.9
- *
- * @return ecma value
- *         Returned value must be freed with ecma_free_value.
- */
-static ecma_value_t
-ecma_builtin_regexp_prototype_symbol_search (ecma_value_t this_arg, /**< this argument */
-                                             ecma_value_t string_arg) /**< string argument */
-{
-  return ecma_regexp_search_helper (this_arg, string_arg);
-} /* ecma_builtin_regexp_prototype_symbol_search */
+#if ENABLED (JERRY_BUILTIN_ANNEXB)
+  if (builtin_routine_id == ECMA_REGEXP_PROTOTYPE_ROUTINE_COMPILE)
+  {
+    return ecma_builtin_regexp_prototype_compile (this_arg, arguments_list_p[0], arguments_list_p[1]);
+  }
+#endif /* ENABLED (JERRY_BUILTIN_ANNEXB) */
 
-/**
- * The RegExp.prototype object's '@@split' routine
- *
- * See also:
- *          ECMA-262 v6.0, 21.2.5.11
- *
- * @return ecma value
- *         Returned value must be freed with ecma_free_value.
- */
-static ecma_value_t
-ecma_builtin_regexp_prototype_symbol_split (ecma_value_t this_arg, /**< this argument */
-                                            ecma_value_t string_arg, /**< source string */
-                                            ecma_value_t limit_arg) /**< limit */
-{
-  return ecma_regexp_split_helper (this_arg, string_arg, limit_arg);
-} /* ecma_builtin_regexp_prototype_symbol_split */
+  /* 1. */
+  if (!ecma_is_value_object (this_arg))
+  {
+    return ecma_raise_type_error (ECMA_ERR_MSG ("'this' is not an object"));
+  }
 
-/**
- * The RegExp.prototype object's '@@match' routine
- *
- * See also:
- *          ECMA-262 v6.0, 21.2.5.6
- *
- * @return ecma value
- *         Returned value must be freed with ecma_free_value.
- */
-static ecma_value_t
-ecma_builtin_regexp_prototype_symbol_match (ecma_value_t this_arg, /**< this argument */
-                                            ecma_value_t string_arg) /**< source string */
-{
-  return ecma_regexp_match_helper (this_arg, string_arg);
-} /* ecma_builtin_regexp_prototype_symbol_match */
+  /* 2. */
+  ecma_object_t *obj_p = ecma_get_object_from_value (this_arg);
+
+#if !ENABLED (JERRY_ESNEXT)
+  bool require_regexp = builtin_routine_id <= ECMA_REGEXP_PROTOTYPE_ROUTINE_TO_STRING;
+#else /* ENABLED (JERRY_ESNEXT) */
+  bool require_regexp = builtin_routine_id < ECMA_REGEXP_PROTOTYPE_ROUTINE_TEST;
 #endif /* ENABLED (JERRY_ESNEXT) */
 
+  if (require_regexp && !ecma_object_class_is (obj_p, LIT_MAGIC_STRING_REGEXP_UL))
```

Comment:
I would prefer to do the RegExp check first, so that error messages remain consistent. Otherwise when we have a non-object this value where a RegExp is required, the error message would only state that it should be an object.